cluster-provision/gocli/pkg/k8s/k8s.go

```diff
@@ -0,0 +1,222 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/dynamic/fake"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/testing"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/yaml"
+)
+
+type K8sDynamicClient interface {
+	Get(gvk schema.GroupVersionKind, name, ns string) (*unstructured.Unstructured, error)
+	Apply(manifest []byte) error
+	List(gvk schema.GroupVersionKind, ns string) (*unstructured.UnstructuredList, error)
+	Delete(gvk schema.GroupVersionKind, name, ns string) error
+}
+
+type K8sDynamicClientImpl struct {
+	scheme *runtime.Scheme
+	client dynamic.Interface
+}
+type ReactorConfig struct {
+	verb      string
+	resource  string
+	reactfunc func(action testing.Action) (bool, runtime.Object, error)
+}
+
+func InitConfig(manifestPath string, apiServerPort uint16) (*rest.Config, error) {
+	config, err := clientcmd.BuildConfigFromFlags("", manifestPath)
+	if err != nil {
+		return nil, fmt.Errorf("Error building kubeconfig: %v", err)
+	}
+	config.Host = "https://127.0.0.1:" + fmt.Sprintf("%d", apiServerPort)
+	config.Insecure = true
+	config.CAData = []byte{}
+	return config, nil
+}
+
+func NewDynamicClient(config *rest.Config) (*K8sDynamicClientImpl, error) {
+	dynamicClient, err := dynamic.NewForConfig(config)
+	if err != nil {
+		return nil, fmt.Errorf("Error creating dynamic client: %v", err)
+	}
+	s := runtime.NewScheme()
+	scheme.AddToScheme(s)
+
+	apiextensionsv1.AddToScheme(s)
+	cephv1.AddToScheme(s)
+	monitoringv1alpha1.AddToScheme(s)
+	monitoringv1.AddToScheme(s)
+	istiov1alpha1.AddToScheme(s)
+	admissionv1.AddToScheme(s)
+	cdiv1beta1.AddToScheme(s)
+	aaqv1alpha1.AddToScheme(s)
+
+	return &K8sDynamicClientImpl{
+		client: dynamicClient,
+		scheme: s,
+	}, nil
+}
+
+func NewTestClient(reactors ...ReactorConfig) *K8sDynamicClientImpl {
+	s := runtime.NewScheme()
+	scheme.AddToScheme(s)
+
+	apiextensionsv1.AddToScheme(s)
+	cephv1.AddToScheme(s)
+	monitoringv1alpha1.AddToScheme(s)
+	monitoringv1.AddToScheme(s)
+	istiov1alpha1.AddToScheme(s)
+	admissionv1.AddToScheme(s)
+	cdiv1beta1.AddToScheme(s)
+	aaqv1alpha1.AddToScheme(s)
+
+	dynamicClient := fake.NewSimpleDynamicClient(s)
+	for _, r := range reactors {
+		dynamicClient.PrependReactor(r.verb, r.resource, r.reactfunc)
+	}
+
+	return &K8sDynamicClientImpl{
+		client: dynamicClient,
+		scheme: s,
+	}
+}
+
+func NewReactorConfig(v string, r string, reactfunc func(action testing.Action) (bool, runtime.Object, error)) ReactorConfig {
+	return ReactorConfig{
+		verb:      v,
+		resource:  r,
+		reactfunc: reactfunc,
+	}
+}
+
+func (c *K8sDynamicClientImpl) Get(gvk schema.GroupVersionKind, name, ns string) (*unstructured.Unstructured, error) {
+	resourceClient, err := c.initResourceClientForGVKAndNamespace(gvk, ns)
+	if err != nil {
+		return nil, err
+	}
+
+	obj, err := resourceClient.Get(context.TODO(), name, v1.GetOptions{})
```

Comment:
while i originally didn't plan for context to be used but the k8s client-go package uses it. so it can be a nice motivation if we later want to adopt context across the repo. for the time being i'll swap it to `context.Background`